scripts/apps/authoring/attachments/attachments.ts

```diff
@@ -0,0 +1,162 @@
+import AttachmentsEditorDirective from './AttachmentsEditorDirective';
+import AttachmentsEditorModalDirective from './AttachmentsEditorModalDirective';
+
+import {
+    closeEdit,
+    selectFiles,
+} from './actions';
+
+import {IAttachment} from '.';
+
+class AttachmentsController {
+    closeEdit: () => void;
+    selectFiles: (files: Array<File>) => void;
+
+    edit: IAttachment;
+    files: Array<IAttachment>;
+    maxSize: number;
+    maxFiles: number;
+    editable: boolean;
+    isLocked: boolean;
+    isLockedByMe: boolean;
+
+    constructor($scope) {
+        const {dispatch} = $scope.store;
+
+        // map dispatch
+        this.selectFiles = (files) => dispatch(selectFiles(files));
+        this.closeEdit = () => dispatch(closeEdit());
+
+        let state = $scope.store.getState();
+
+        // re-render on state change
+        const unsubscribe = $scope.store.subscribe(() => {
+            if ($scope.store.getState().attachments !== state.attachments) {
+                state = $scope.store.getState();
+                $scope.$applyAsync(() => {
+                    this.mapState($scope.store.getState());
+                });
+            }
+        });
+
+        $scope.$on('$destroy', unsubscribe);
+
+        // init
+        this.mapState($scope.store.getState());
+    }
+
+    mapState(state) {
+        this.edit = state.attachments.edit;
+        this.files = state.attachments.files;
+        this.maxSize = state.attachments.maxSize;
+        this.maxFiles = state.attachments.maxFiles;
+        this.editable = state.editor.editable;
+        this.isLocked = state.editor.isLocked;
+        this.isLockedByMe = state.editor.isLockedByMe;
+    }
+}
+
+AttachmentsController.$inject = ['$scope'];
+
+AttachmentsFactory.$inject = ['api'];
+function AttachmentsFactory(api) {
+    const RESOURCE = 'attachments';
+    const _byIdCache = {};
+
+    class AttachmentsService {
+        /**
+         * Get attachments by item
+         *
+         * @param {Object} item
+         * @return {Promise}
+         */
+        byItem(item) {
+            const attachments = item.attachments || [];
+
+            return api.query(RESOURCE, {where: {_id: {$in: attachments.map((ref) => ref.attachment)}}})
+                .then((data) => data._items);
+        }
+
+        /**
+         * Save attachment changes
+         *
+         * @param {Object} file
+         * @param {Object} diff
+         * @return {Promise}
+         */
+        save(file, diff) {
+            clearCache(file);
+            return api.save(RESOURCE, file, diff);
+        }
+
+        /**
+         * Remove attachment
+         *
+         * @param {Object} file
+         * @return {Promise}
+         */
+        remove(file) {
+            clearCache(file);
+            return api.remove(file);
+        }
+
+        /**
+         * Get attachment by id
+         *
+         * @param {String} id
+         * @return {Promise}
+         */
+        byId(id) {
```

Comment:
Is the method used anywhere? I didn't see it in current changes. Does the cache live through application lifecycle and isn't specific to any view? How are cached items updated when someone else changes them?

Reply:
I think it's used in editor when rendering attachment used as link